libs/ktem/ktem/index/file/pipelines.py

```diff
@@ -236,226 +249,356 @@ def get_pipeline(cls, user_settings, index_settings, selected):
+                    f"File {file_path.name} already indexed. Please rerun with "
+                    "reindex=True to force reindexing."
+                )
+            else:
+                # remove the existing records
+                with Session(engine) as session:
+                    session.execute(
+                        delete(self.Source).where(self.Source.id == file_id)
+                    )
+                    vs_ids, ds_ids = [], []
+                    index = session.execute(
+                        select(self.Index).where(self.Index.source_id == file_id)
+                    ).all()
+                    for each in index:
+                        if each[0].relation_type == "vector":
+                            vs_ids.append(each[0].target_id)
+                        else:
+                            ds_ids.append(each[0].target_id)
+                        session.delete(each[0])
+                    session.commit()
+                self.VS.delete(vs_ids)
+                self.DS.delete(ds_ids)
+                file_id = self.store_file(file_path)
+        else:
+            # add record to db
+            file_id = self.store_file(file_path)
+
+        # extract the file
+        extra_info = default_file_metadata_func(str(file_path))
+        docs = self.loader.load_data(file_path, extra_info=extra_info)
+        for _ in self.handle_docs(docs, file_id, file_path.name):
+            continue
+        self.finish(file_id, file_path)
+
+        return file_id
+
+    def stream(
+        self, file_path: str | Path, reindex: bool, **kwargs
+    ) -> Generator[Document, None, str]:
+        # check for duplication
+        file_path = Path(file_path).resolve()
+        file_id = self.get_id_if_exists(file_path)
+        if file_id is not None:
+            if not reindex:
+                raise ValueError(
+                    f"File {file_path.name} already indexed. Please rerun with "
+                    "reindex=True to force reindexing."
+                )
+            else:
+                # remove the existing records
+                yield Document(f" => Removing old {file_path.name}", channel="debug")
+                with Session(engine) as session:
+                    session.execute(
+                        delete(self.Source).where(self.Source.id == file_id)
+                    )
+                    vs_ids, ds_ids = [], []
+                    index = session.execute(
+                        select(self.Index).where(self.Index.source_id == file_id)
+                    ).all()
+                    for each in index:
+                        if each[0].relation_type == "vector":
+                            vs_ids.append(each[0].target_id)
+                        else:
+                            ds_ids.append(each[0].target_id)
+                        session.delete(each[0])
+                    session.commit()
+                self.VS.delete(vs_ids)
+                self.DS.delete(ds_ids)
```

Comment:
same as above

Reply:
Changed in [bfff3c9](https://github.com/Cinnamon/kotaemon/pull/81/commits/bfff3c9a5aa1be55d4b79caff046a3f5858c9d3d)